### PR TITLE
Allow serving media through a redirect to an unauthenticated signed URL

### PR DESCRIPTION
**Pull request by sandhose**
This can be reviewed commit by commit.

The general idea is to make media requests (both C-S and S-S) redirect to a new set of endpoints at `/_synapse/media/*`. The URLs are signed and valid for a short time (10 minutes by default).

The reasons for doing this are multiple:

 - the URLs serving media are now shared between client and federation requests, meaning a CDN can use a common cache for them
 - the federation media requests now serve the media outside of a multipart/mixed body, allowing them to go through file analysis tools
 - we now have the ability to generate abritrary short-lived URLs for any media without a full access token or federation authorization, which will be useful for displaying user avatars in MAS
 - we could now imagine serving all the medias on a dedicated domain, making it clear that the domain serves user-generated content

All of this is allowed by the spec, but I had to do one concession: initially the federation media location was supposed to be relative (e.g. `/_synapse/media/…`), but Synapse has a bug where when it downloads media through federation, it will interpret a redirect as absolute. This means I had to make the redirect absolute (e.g. `https://example.com/_synapse/media/…`).

This is not enabled by default, as it requires routing the `/_synapse/media` URLs to the media worker and setting up an extra secret.

Internal reference: https://github.com/element-hq/backend-internal/issues/144#issuecomment-3158368107

### CODE REVIEW

File: schema/synapse-config.schema.yaml [hunk 1]
```diff
@@ -2433,6 +2433,44 @@ properties:
           max_size: 100M
         - time_period: 1w
           max_size: 500M
+  media_redirect:
+    type: object
+    description: >-
+      When enabled, Synapse will use HTTP redirect responses to serve media
+      instead of directly serving the media from the media store. This can help
+      with caching, but requires /_synapse/media/* to be routed to the media
```
Comment (MadLittleMods) [inline schema/synapse-config.schema.yaml:2441]:
```suggestion
      with caching, but requires `/_synapse/media/*` to be routed to the media
```

File: docker/configure_workers_and_start.py [hunk 2]
```diff
@@ -127,6 +127,7 @@
             "^/_synapse/admin/v1/quarantine_media/.*$",
             "^/_matrix/client/v1/media/.*$",
             "^/_matrix/federation/v1/media/.*$",
+            "^/_synapse/media/",
```
Comment (MadLittleMods) [inline docker/configure_workers_and_start.py:130]: Should we just put this next to `^/_matrix/media/` at the top?

File: schema/synapse-config.schema.yaml (hunk 1 above)
Comment (MadLittleMods) [inline schema/synapse-config.schema.yaml:2441]: Overall, I feel like this description doesn't paint a full enough picture of what this does or why I should care.

The `/_matrix/media/` URL's redirect `/_synapse/media/` which serves the media directly?

Additionally, I have no idea how to set this up with Cloudflare for example.

File: schema/synapse-config.schema.yaml [hunk 3]
```diff
@@ -2433,6 +2433,44 @@ properties:
...
+      secret_path:
+        type: ["string", "null"]
+        description: >-
+          An alternative to `media_redirect.secret` that specifies a file
+          containing the secret.
+        default: null
+      ttl:
+        $ref: "#/$defs/duration"
+        description: >-
+          How long the redirect URLs should be valid for.
```
Comment (MadLittleMods) [inline schema/synapse-config.schema.yaml:2465]: What are the considerations for long this should be set for? (update the docs with the answer)

This only needs long enough for the client to request the media and the CDN to cache it?

File: schema/synapse-config.schema.yaml (hunk 1 above)
Comment (MadLittleMods) [inline schema/synapse-config.schema.yaml:2441]: For my own benefit, can you explain the cache benefits please?

Is it only that CS and SS API's serve from the same source?

Based on https://github.com/element-hq/backend-internal/issues/144, I'm guessing there is more to this but my eyes are glazing over trying to glean anything from that issue.

File: synapse/rest/synapse/media/download.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                sig,
+            )
+            raise NotFoundError()
+
+        # Check the expiry time
+        if exp < self._clock.time_msec():
+            logger.info("Expired signed URL serving media %s", media_id)
+            raise NotFoundError()
+
+        # Reply with the media
```
Comment (MadLittleMods) [inline synapse/rest/synapse/media/download.py:103]:
```suggestion
        # Reply with the media directly
```

File: synapse/rest/synapse/media/thumbnail.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            raise NotFoundError()
+
+        # Now parse and check the parameters
+        args = parse_qs(base64.urlsafe_b64decode(parameters))
+        width = parse_integer_from_args(args, "width", required=True)
+        height = parse_integer_from_args(args, "height", required=True)
+        method = parse_string_from_args(args, "method", "scale")
+        m_type = parse_string_from_args(args, "type", "image/png")
+
+        # Reply with the thumbnail
```
Comment (MadLittleMods) [inline synapse/rest/synapse/media/thumbnail.py:122]:
```suggestion
        # Reply with the thumbnail directly
```

File: synapse/rest/synapse/media/download.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        await self._media_repository.get_local_media(
+            request=request,
+            media_id=media_id,
+            name=name,
+            max_timeout_ms=0,  # If we got here, the media finished uploading
+            federation=False,  # This changes the response to be multipart; we explicitly don't want that
```
Comment (MadLittleMods) [inline synapse/rest/synapse/media/download.py:109]: We should explain why

This allows them to go through file analysis tools such as ...

File: synapse/rest/synapse/media/download.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        if exp < self._clock.time_msec():
+            logger.info("Expired signed URL serving media %s", media_id)
+            raise NotFoundError()
+
+        # Reply with the media
+        await self._media_repository.get_local_media(
+            request=request,
+            media_id=media_id,
+            name=name,
+            max_timeout_ms=0,  # If we got here, the media finished uploading
```
Comment (MadLittleMods) [inline synapse/rest/synapse/media/download.py:108]: Not specific to this PR but since this timeout option only applies to the media upload, we should probably just rename it

File: synapse/rest/synapse/media/download.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            federation=False,  # This changes the response to be multipart; we explicitly don't want that
+            may_redirect=False,  # We're already on the redirected URL, we don't want to redirect again
```
Comment (MadLittleMods) [inline synapse/rest/synapse/media/download.py:110]:
```suggestion
            may_redirect=False,  # We're already on the redirected URL, we don't want to redirect again; serve the media directly
```

Comment (MadLittleMods) [inline synapse/rest/synapse/media/download.py:110]: nit: Putting the comments above the options would be my preference :shrug: 

Feel free to keep your way. We don't have a lint for this.

File: synapse/media/media_repository.py [hunk 9]
```diff
@@ -1545,3 +1571,145 @@ async def _remove_local_media_from_disk(
...
+    ) -> str:
+        """Get the signed location for a media download
+
+        That URL will serve the media with no extra authentication for a limited
+        time, allowing the media to be cached by a CDN more easily.
+        """
+
+        # XXX: One potential improvement here would be to round the `exp` to
+        # the nearest 5 minutes, so that a CDN/cache can always cache the
+        # media for a little bit
```
Comment (MadLittleMods) [inline synapse/media/media_repository.py:1608]: Why is that better than configurable `redirect_ttl_ms`?

File: synapse/media/media_repository.py [hunk 10]
```diff
@@ -1545,3 +1571,145 @@ async def _remove_local_media_from_disk(
...
+        name: Optional[str] = None,
+    ) -> StrSequence:
+        """Get the key used for the download media signature
+
+        Args:
+            media_id: The media ID of the content. (This is the same as
+                the file_id for local content.)
+            exp: The expiration time of the signature, as a unix timestamp in ms.
+            name: Optional name that, if specified, will be used as
+                the filename in the Content-Disposition header of the response.
```
Comment (MadLittleMods) [inline synapse/media/media_repository.py:1588]: How does that work?

Where do we go from parsing the key to putting it in the `Content-Disposition` header?

I do see that we have a test that may be verifying this behavior.

File: synapse/media/media_repository.py [hunk 11]
```diff
@@ -1545,3 +1571,145 @@ async def _remove_local_media_from_disk(
...
+            parameters: The parameters of the thumbnail request as a string,
+                e.g. "width=100&height=100"
+            exp: The expiration time of the signature, as a unix timestamp in ms.
+        """
+        return ("thumbnail", media_id, parameters, str(exp))
+
+    def signed_location_for_thumbnail(
+        self,
+        media_id: str,
+        parameters: dict[str, str],
```
Comment (MadLittleMods) [inline synapse/media/media_repository.py:1642]: We should explain what should go in the `parameters` -> Looks like we explain above in `thumbnail_media_key(...)`. If we want to avoid the duplication, we should at-least point to it.

Perhaps this should be a `TypedDict` with known keys

File: schema/synapse-config.schema.yaml (hunk 3 above)
Comment (MadLittleMods) [inline schema/synapse-config.schema.yaml:2465]: What was the point of authenticated media?

I'm assuming it was so that people can't view media without having room membership?

If we're allowing a CDN to cache media, won't the CDN URL just serve the media into the future. Do we invalidate the CDN ever?

File: tests/federation/test_federation_media.py [hunk 12]
```diff
@@ -186,6 +186,72 @@ def test_federation_etag(self) -> None:
...
+        self.assertRegex(
+            lines[5],
+            rf"^Location: https://test/_synapse/media/download/{content_uri.media_id}\?exp=\d+&sig=\w+$",
+        )
+        self.assertEqual(lines[6], "")  # First empty line after Location header
+        self.assertEqual(lines[7], "")  # Second empty line after Location header
+
+        # Final boundary
+        self.assertEqual(lines[8], f"--{boundary}--")
+        self.assertEqual(lines[9], "")
```
Comment (MadLittleMods) [inline tests/federation/test_federation_media.py:253]: While these assertions work, they don't provide any context when they fail.

It would be nice to do something like this:

```python
assertStringSequenceMatches(
    channel.text_body.split("\r\n"),
    [
        f"--{boundary}",
        "Content-Type: application/json",
        "",
        "{}",
        f"--{boundary}",
        re.compile(rf"^Location: https://test/_synapse/media/download/{content_uri.media_id}\?exp=\d+&sig=\w+$"),
        "",
        "",
        f"--{boundary}",
        "",
    ],
)
```

File: tests/federation/test_federation_media.py [hunk 13]
```diff
@@ -186,6 +186,72 @@ def test_federation_etag(self) -> None:
...
+                46,
+                UserID.from_string("@user_id:whatever.org"),
+            )
+        )
+        # test with a text file
+        channel = self.make_signed_federation_request(
+            "GET",
+            f"/_matrix/federation/v1/media/download/{content_uri.media_id}",
+        )
+        self.pump()
```
Comment (MadLittleMods) [inline tests/federation/test_federation_media.py:209]: Test seems to pass without this

File: tests/federation/test_federation_media.py [hunk 14]
```diff
@@ -186,6 +186,72 @@ def test_federation_etag(self) -> None:
...
+        self.assertEqual(200, channel.code)
+
+        content_type = channel.headers.getRawHeaders("content-type")
+        assert content_type is not None
+        assert "multipart/mixed" in content_type[0]
+        assert "boundary" in content_type[0]
```
Comment (MadLittleMods) [inline tests/federation/test_federation_media.py:215]: :shrug: Since the syntax of this header matters, we could use this:

```suggestion
        content_type = channel.headers.getRawHeaders("content-type")
        assert content_type is not None
        assert content_type[0].startswith("multipart/mixed; boundary=")
```

Since we extract the boundary value anyway, perhaps we should just go for a regex with a capture group though

File: tests/rest/client/test_media.py [hunk 15]
```diff
@@ -2527,6 +2527,84 @@ def test_same_quality(self, method: str, desired_size: int) -> None:
...
+        # Request the download
+        channel = self.make_request(
+            "GET",
+            f"/_matrix/client/v1/media/download/{self.hs.hostname}/{content_uri.media_id}",
+            shorthand=False,
+            access_token=self.tok,
+        )
+
+        # Should get a redirect response
+        self.assertEqual(channel.code, 307)
```
Comment (MadLittleMods) [inline tests/rest/client/test_media.py:2554]:
```suggestion
        self.assertEqual(channel.code, HTTPStatus.TEMPORARY_REDIRECT)
```

File: tests/federation/test_federation_media.py [hunk 16]
```diff
@@ -293,3 +359,69 @@ def test_thumbnail_download_cropped(self) -> None:
...
+        )
+        # test with a thumbnail request
+        channel = self.make_signed_federation_request(
+            "GET",
+            f"/_matrix/federation/v1/media/thumbnail/{content_uri.media_id}?width=32&height=32&method=scale",
+        )
+        self.pump()
+        self.assertEqual(200, channel.code)
+
+        content_type = channel.headers.getRawHeaders("content-type")
```
Comment (MadLittleMods) [inline tests/federation/test_federation_media.py:386]: Doesn't seem to matter but to match other spots we do this:

```suggestion
        content_type = channel.headers.getRawHeaders("Content-Type")
```

Comment (MadLittleMods) [inline tests/federation/test_federation_media.py:386]: We could assert a single header like we do here:

https://github.com/element-hq/synapse/blob/be5ce0c6b54bdd0d1b9bf1d169f705fe1316f4c4/tests/rest/client/test_media.py#L2556-L2561

File: tests/rest/synapse/media/test_download.py [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+                "text/plain",
+                "some_name.txt",
+                content,
+                17,
+                UserID.from_string("@user:test"),
+            )
+        )
+
+        # Generate a signed URL
+        exp = self.clock.time_msec() + 3600000  # 1 hour from now
```
Comment (MadLittleMods) [inline tests/rest/synapse/media/test_download.py:64]: We might as well use `synapse/util/constants.py` for clarity and avoiding mistakes.

```suggestion
        exp = self.clock.time_msec() + ONE_HOUR_MS
```

File: tests/rest/synapse/media/test_thumbnail.py [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+        # Create test content with an image
+        content = io.BytesIO(small_png.data)
+        content_uri = self.get_success(
+            self.media_repo.create_or_update_content(
+                "image/png",
+                "test_image.png",
+                content,
+                67,
+                UserID.from_string("@user:test"),
+            )
```
Comment (MadLittleMods) [inline tests/rest/synapse/media/test_thumbnail.py:111]: Using keyword args might make this more clear. I had to look what the `67` was for.

Seems good especially as we have two string args next to each other that could easily be mixed up. We can use `*` to force keyword arg usage

File: tests/rest/synapse/media/test_thumbnail.py [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+    def test_valid_signed_thumbnail_cropped(self) -> None:
+        """Test that a valid signed URL returns the thumbnail content (cropped)"""
+        # Create test content with an image
+        content = io.BytesIO(small_png.data)
+        content_uri = self.get_success(
+            self.media_repo.create_or_update_content(
+                "image/png",
+                "test_image.png",
+                content,
+                67,
```
Comment (MadLittleMods) [inline tests/rest/synapse/media/test_thumbnail.py:109]:
```suggestion
                len(SMALL_PNG),
```

File: tests/rest/synapse/media/test_thumbnail.py [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+            shorthand=False,
+        )
+        self.pump()
+
+        # Check the response
+        self.assertEqual(channel.code, 200)
+
+        # Advance the clock to make the URL expired
+        self.reactor.advance(
+            10 * 60 + 1
```
Comment (MadLittleMods) [inline tests/rest/synapse/media/test_thumbnail.py:224]:
```suggestion
            10 * ONE_MINUTE_SECONDS + 1
```
